File: ICPC NAQ F23/B.py

word = input()


def isnumber(word):
    
     
    if word == "zero":
        return True
    elif word == "one":
        return True
    elif word ==  "two":
        return True
    elif word ==  "three":
        return True
    elif word ==  "four":
        return True
    elif word ==  "five":
        return True
    elif word == "six":
        return True
    elif word ==  "seven":
        return True
    elif word == "eight":
        return True
    elif word ==  "nine":
        return True
    else:
        return False

columns = len(word) + 1
rows = int(columns / 3) + 2



results = []
uniqueness = [1]

for i in range(rows):
    results.append([])
    uniqueness.append(1)
    for j in range(columns):
        if (i == 0):
            results[i].append(j)
        else:
            if (j == 0):
                results[i].append(0)
            else:
                subword = word[:j]
                threeword = subword[-3:]
                fourword = subword[-4:]
                fiveword = subword [-5:]
                if (isnumber(threeword)):
                    if (j - 3 >= 0):
                        previous_best = results[i-1][j-3]
                        if (previous_best < results[i][j-1]):
                            results[i].append( previous_best + 1)
                        elif (previous_best == results[i][j-1]):
                            results[i].append( previous_best + 1)
                            uniqueness[i] += 1
                        else:
                            results[i].append (results[i][j-1] + 1)
                elif (isnumber(fourword)):
                    if (j - 4 >= 0):
                        previous_best = results[i-1][j-4]
                        if (previous_best < results[i][j-1]):
                            results[i].append( previous_best + 1)
                        elif (previous_best == results[i][j-1]):
                            results[i].append( previous_best + 1)
                            uniqueness[i] += 1
                        else:
                            results[i].append (results[i][j-1] + 1)
                elif(isnumber(fiveword)):
                    if (j - 5 >= 0):
                        previous_best = results[i-1][j-5]
                        if (previous_best < results[i][j-1]):
                            results[i].append( previous_best + 1)
                        elif (previous_best == results[i][j-1]):
                            results[i].append( previous_best + 1)
                            uniqueness[i] += 1
                        else:
                            results[i].append (results[i][j-1] + 1)
                else:
                    results[i].append(results[i][j-1] + 1)
    #print(results[i])

print(results[i][j])
print(uniqueness[i])
    

File: ICPC NAQ F23/B.cpp
#include <iostream>
#include <vector>
#include <string>

bool isNumber(const std::string& word) {
    if (word == "zero" || word == "one" || word == "two" || word == "three" ||
        word == "four" || word == "five" || word == "six" || word == "seven" ||
        word == "eight" || word == "nine") {
        return true;
    }
    return false;
}

int main() {
    std::string word;
    std::cin >> word;

    int columns = word.length() + 1;
    int rows = columns / 3 + 2;

    std::vector<std::vector<int>> results(rows, std::vector<int>(columns, 0));
    std::vector<int> uniqueness(rows, 1);

    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < columns; ++j) {
            if (i == 0) {
                results[i][j] = j;
            } else {
                if (j == 0) {
                    results[i][j] = 0;
                } else {
                    std::string subword = word.substr(0, j);
                    std::string threeword = subword.substr(std::max(0, j - 3), 3);
                    std::string fourword = subword.substr(std::max(0, j - 4), 4);
                    std::string fiveword = subword.substr(std::max(0, j - 5), 5);

                    if (isNumber(threeword)) {
                        if (j - 3 >= 0) {
                            int previous_best = results[i - 1][j - 3];
                            if (previous_best < results[i][j - 1]) {
                                results[i][j] = previous_best + 1;
                            } else if (previous_best == results[i][j - 1]) {
                                results[i][j] = previous_best + 1;
                                uniqueness[i]++;
                            } else {
                                results[i][j] = results[i][j - 1] + 1;
                            }
                        }
                    } else if (isNumber(fourword)) {
                        if (j - 4 >= 0) {
                            int previous_best = results[i - 1][j - 4];
                            if (previous_best < results[i][j - 1]) {
                                results[i][j] = previous_best + 1;
                            } else if (previous_best == results[i][j - 1]) {
                                results[i][j] = previous_best + 1;
                                uniqueness[i]++;
                            } else {
                                results[i][j] = results[i][j - 1] + 1;
                            }
                        }
                    } else if (isNumber(fiveword)) {
                        if (j - 5 >= 0) {
                            int previous_best = results[i - 1][j - 5];
                            if (previous_best < results[i][j - 1]) {
                                results[i][j] = previous_best + 1;
                            } else if (previous_best == results[i][j - 1]) {
                                results[i][j] = previous_best + 1;
                                uniqueness[i]++;
                            } else {
                                results[i][j] = results[i][j - 1] + 1;
                            }
                        }
                    } else {
                        results[i][j] = results[i][j - 1] + 1;
                    }
                }
            }
        }
    }

    std::cout << results[rows - 1][columns - 1] << std::endl;
    std::cout << uniqueness[rows - 1] << std::endl;

    return 0;
}
